### DISCUSSION TITLE
Pyinstaller binary launching browser and libstdc++.so.6

### DISCUSSION THREAD

**Discussion by varun-dc** [pyinstaller/pyinstaller#6480] [PyInstaller] [answered]
Hi,

I've got a project, commercial/proprietary, so I'm not able to share sources. But I've got a question regarding a particular situation where,
we're using `dbt` (https://github.com/dbt-labs/dbt-core/) which is using snowflake-connector-python (https://github.com/snowflakedb/snowflake-connector-python) which is triggering the user's browser to open (using `import webbrowser` and `webbrowser.open_new(...)`). This is on Linux.

We're compiling this python program, which contains code that does what I've mentioned above, into a standalone binary. What we're seeing is, the end user is getting an error when the snowflake-connector-python is triggering the browser to open. The error appears to be on occurring in the browser itself, not our code or the pyinstaller binary. Which makes sense as it's only doing a system call to open the browser.

>     /usr/lib/firefox/firefox:
>      /home/user/.config/Electron/dbt-subsys-tmp/_MEIntWdeX/libstdc++.so.6:
>         version `GLIBCXX_3.4.26' not found (required by /usr/lib/firefox/firefox)

Here `/home/user/.config/Electron/dbt-subsys-tmp/_MEIntWdeX` is the temp dir created by pyinstaller when our binary unpacks and executes.

I wanted to see if libstdc++.so.6 is included in the packed binary. It is included. But the file's `md5sum` hash doesn't match the distro's version of the file. We're compiling the pyinstaller binary on a separate system as part of our build process, so there's no guarantee we're going to be shipping a identical binary .so.6 file the user may have installed on their system.

One way I've found to work around this issue is by noticing pyinstaller sets the `LD_LIBRARY_PATH`, so I at runtime prepend `/usr/lib/x86_64-linux-gnu/` to that variable before the rest of the code runs to run the browser. This fixes the issue, but for a while we weren't sure why this is actually fixing things.

Now I've realized what's possibly happening, and why the fix above is actually working: because the browser (chrome, firefox, etc) is expecting a specific libstdc++, which the user's Linux distro has installed. But since we're bundling a specific, likely a different minor version of libstdc++.so, in our pyinstaller binary, and pyinstaller is setting the `LD_LIBRARY_PATH`, the browser is not finding the expected distro's version of libstdc++.so.6.

So my current theory is, this fix I've described above is the right approach for this situation. The users browser should be allowed to use the Linux distro provided libstdc++, not the one pyinstaller providers. So we should either prepend to `LD_LIBRARY_PATH` as we're doing (or possible use `LD_PRELOAD`?). The only possible downside here is we may end up causing something else that depends on libstdc++ in the pyinstaller bundle to now also use the user's libstdc++.so instead of the pyinstaller version.

Alternatively I suppose we can monkey patch the `webbrowser.open_new` function, restoring and reverting the LD_LIBRARY_PATH env var before and after the call into the real `open_new` function.

But maybe this isn't an issue as long as we ensure we build the pyinstaller binary on the older system we wish to support. I was having a read through some of the discussions here, https://stackoverflow.com/questions/11107263/how-compatible-are-different-versions-of-glibc, and it seems like as long we compile against an other version, newer versions should work due to backward compatibility, but not the other way around as there is no forward compatibility.

I was wondering if anyone more experienced with pyinstaller and dynamic libraries could verify my reasoning here and suggest if this is a good approach or not?

Thanks

**Comment by rokm** [comment, marked as answer]
See #6334 for the same sort of discussion.

But yes, restoring/reverting `LD_LIBRARY_PATH` is the proper (and [documented](https://pyinstaller.readthedocs.io/en/stable/runtime-information.html#ld-library-path-libpath-considerations)) approach. 

You can also restore/revert `LD_LIBRARY_PATH` at the start of your program (e.g., in the frozen-application-specific codepath), if you are certain that no 3rd party helper executables were collected into the frozen bundle that would require the PyInstaller-specific `LD_LIBRARY_PATH` when spawned (e.g., glib's spawn helper).
